README
Chat Socks

Titanium Appcelerator implementation of socket.io chat example

Socket.io Chat Example 
https://github.com/LearnBoost/socket.io/tree/master/examples/chat

Using a webview and Global events.

----------------------------------
License

(The MIT License)

Copyright (c) 2011 Christian Sullivan <christian@xogix.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.txt
v0.0.1
Hello World

Resources/app.js
Ti.include('/main.js');
Titanium.UI.setBackgroundColor('#000');
var win = Ti.UI.createWindow();

socket.connect(win);
/*
 * Header user list
 */
var header = Titanium.UI.createScrollView({
	contentWidth:'100%',
	contentHeight:50,
	top:0,
	left:0,
	showHorizontalScrollIndicator:true
});
var headerContainer= Ti.UI.createView({
	top:0,
	left:0,
	width:'100%',
	height:50,
	color:'#999999'
})
headerContainer.add(header)
win.add(headerContainer);

/*
 * Message list container
 */
var container = Ti.UI.createTableView({
	top:50,
	left:0,
	width:320,
	height:370,
});

win.add(container);

/*
 * Input Toolbar
 */
var flexSpace = Titanium.UI.createButton({
	systemButton:Titanium.UI.iPhone.SystemButton.FLEXIBLE_SPACE
});
var textfield = Titanium.UI.createTextField({
	height:32,
	backgroundImage:'images/inputfield.png',
	width:200,
	font: {
		fontSize:13
	},
	color:'#777',
	paddingLeft:10,
	borderStyle:Titanium.UI.INPUT_BORDERSTYLE_NONE,
	returnKeyType:Titanium.UI.RETURNKEY_SEND
});
var sendButton = Titanium.UI.createButton({
	backgroundImage:'images/send.png',
	backgroundSelectedImage:'images/send_selected.png',
	width:67,
	height:32
});
var toolbar = Titanium.UI.createToolbar({
	items:[flexSpace,textfield,flexSpace, sendButton,flexSpace],
	bottom:0,
	borderTop:true,
	borderBottom:false,
	translucent:true,
	barColor:'#999'
});
win.add(toolbar);
toolbar.hide();

var button = Ti.UI.createButton({
	top:-10,
	left:-10,
	height:1,
	width:1,
});
win.add(button);

var userPrompt = xg.ui.userPrompt();
win.add(userPrompt);
userPrompt.top = 50;
userPrompt.height = 150;
userPrompt.width = 310;

/*
 * Event Listeners
 */
socket.on('userlist:set', function(e) {
	//TODO Remove only the user that goes offline instead of complete re-draw
	for(var c in header.children) {
		header.remove(header.children[c]);
	}
	for (var x =0; x <= e.users.length-1; x = x + 1) {
		var currUser = e.users[x];
		var user = [];
		user[currUser] = Ti.UI.createButton({
			title:currUser,
			top:5,
			left:(x*60)+5,
			height:40,
			width:60
		});
		header.add(user[currUser]);
	}
});
socket.on('message:add', function(e) {
	container.appendRow(xg.ui.chatRow(e));
	//TODO Fix Message Scroll issue only fires sometime for incoming messages
	container.scrollToIndex(container.data[0].rowCount);
});
sendButton.addEventListener('click', function() {
	socket.emit('message:out', {
		message:textfield.value
	});
	textfield.value = '';
});
textfield.addEventListener('return', function () {
	socket.emit('message:out', {
		message:textfield.value
	});
	textfield.value = '';
});
textfield.addEventListener('focus', function () {
	container.height = 150;
	toolbar.bottom = 216;
});
textfield.addEventListener('blur', function () {
	container.height = 370;
	toolbar.bottom = 0;
});
socket.on('nickname:get', function(e) {
	Ti.API.info('fired -> nickname:get');
	if(!e.set) {
		userPrompt.fireEvent('show');
		toolbar.hide();
	} else {
		userPrompt.fireEvent('hide');
		toolbar.show();
	}
});
win.open();

Resources/webview/webview.js
var socket = io.connect('http://localhost:3012');
socket.on('connect', function () {
	//Fires event listern in app.js to open the nickname prompt
	Ti.App.fireEvent('nickname:get');
});
function message (from, msg) {
	Ti.App.fireEvent('message:add',{user:from,message:msg});
}
socket.on('announcement', function (msg) {
	message('Announcement',msg);
});
socket.on('nicknames', function (nicknames) {
	var nicks = [];
	for (var i in nicknames) {
		if(nicknames[i] !== null){
			nicks.push(nicknames[i]);
		}
	}
	Ti.App.fireEvent('userlist:set',{users:nicks});
});
socket.on('reconnect', function () {
	message('System', 'Reconnected to the server');
});
socket.on('reconnecting', function () {
	message('System', 'Attempting to re-connect to the server');
});
socket.on('error', function (e) {
	message('System', e ? e : 'A unknown error occurred');
});
socket.on('user message', message);

/*
 * Event Listeners
 */
(function () {
	//Sets nickname for users socket connection on server
	Ti.App.addEventListener('nickname:set', function (e) {
		socket.emit('nickname', e.user, function (usertaken) {
			if(usertaken) {
				Ti.App.fireEvent('nickname:get', {
					user: e.user,
					set: false,
					error:true
				});
			} else {
				Ti.App.fireEvent('nickname:get', {
					user: e.user,
					set: true,
					error:false
				});

			}
		});
	});
	//Work around for scroll to bottom issue in chat window
	Ti.App.addEventListener('message:scroll', function () {
		Ti.App.fireEvent('messagelist:fix');
	});
	//Sends new message to to server and appends to users chat window
	Ti.App.addEventListener('message:out', function (e) {
		message('me', e.message);
		socket.emit('user message', e.message, function (e) {
		});
	});
})();

Resources/ui.js
xg.ui = {};

/*
 * User Prompt to set nickname
 * Adapted from Dan Tamas's <hi@dan-tamas.me> nifty tooltip
 */
xg.ui.userPrompt = function (trigger) {
	var	container =  Titanium.UI.createView({
		width:'auto',
		height:'auto',
		backgroundImage:"images/bubble.png",
		backgroundTopCap:30.0,
		backgroundLeftCap:10.0,
		top:0,
		right:3,
		opacity:0,
		open:false
	});

	var label = Ti.UI.createLabel({
		text:'Enter Your Nickname',
		color:'#ffffff',
		opacity:1,
		top:30,
		left:10,
		right:10,
		height:'auto',
		width:'auto'
	});

	container.add(label);

	var input = Ti.UI.createTextField({
		top:60,
		width: 150,
		height: 30,
		left:20,
		right:20,
		borderStyle:Titanium.UI.INPUT_BORDERSTYLE_ROUNDED,
		returnKeyType:Titanium.UI.RETURNKEY_DONE
	});
	container.add(input);

	input.addEventListener('return', function () {
		if(input.value !== '') {
			Ti.App.fireEvent('nickname:set', {
				user:input.value
			});
		}
	});
	var anim_out = Titanium.UI.createAnimation();
	anim_out.opacity = 0;
	anim_out.duration = 250;

	var anim_in = Titanium.UI.createAnimation();
	anim_in.opacity = 1;
	anim_in.duration = 250;

	container.addEventListener('show', function() {
		container.animate(anim_in);
	});
	container.addEventListener('hide', function() {
		container.animate(anim_out);
	});
	return container;
}

/*
 * Creates chat row for incoming messages
 */
xg.ui.chatRow = function (msg) {
	var row = Titanium.UI.createTableViewRow({height:'auto'});

	var userLBL =  Ti.UI.createLabel({
		text:msg.user,
		width:'auto',
		height:20,
		font:{
			fontSize:14,
			fontWeight:'bold'
		},
		top:5
	});

	var messageLBL = Titanium.UI.createLabel({
		text:msg.message,
		font: {
			fontSize:12,
		},
		width:'auto',
		top:20,
		bottom:5,
		width:'95%',
		height:'auto'
	});
	if(msg.user == 'me') {
		messageLBL.textAlign = 'right';
		userLBL.right = 5;
	}
	else{
		messageLBL.textAlign = 'left';
		userLBL.left = 5;
	}
	row.add(userLBL);
	row.add(messageLBL);
	row.className = 'message_row';
	return row;
}